meson.build

```diff
@@ -1642,6 +1642,16 @@ update_hwdb_autosuspend_sh = find_program('tools/update-hwdb-autosuspend.sh')
 update_syscall_tables_sh = find_program('tools/update-syscall-tables.sh')
 xml_helper_py = find_program('tools/xml_helper.py')
 
+############################################################
+
+want_rust = get_option('build-rust')
+rustc = find_program('rustc', required : want_rust == 'true')
```

Comment:
I think it should take the `RUSTC` environment variable into account so that it would be possible to replace `rustc` with [`clippy-driver`](https://github.com/rust-lang/rust-clippy#as-a-rustc-replacement-clippy-driver) (or anything else). To judge from https://github.com/mesonbuild/meson/blob/master/docs/markdown/Reference-tables.md#compiler-and-linker-flag-environment-variables, that's what `meson` expects too.

Reply:
So this is used as a dependency check - to see if there is a compiler. But it is not actually used directly to compile - that is left to Meson to implement, via the `add_languages('rust')` mechanism. So, I think it should already work as intended?

Reply:
I tried to pass `RUSTC=clippy-driver` and it failed with
```
Program rustc found: YES (/home/vagrant/.cargo/bin/rustc)
Compiler for language rust for the build machine not found.

meson.build:1652:8: ERROR: Unknown compiler(s): [['clippy-driver']]

A full log can be found at /home/vagrant/systemd/build/meson-logs/meson-log.txt
```
so it seems it doesn't work as expected unfortunately (due to some meson bug probably)

FWIW clippy isn't happy
```
warning: unsafe function's docs miss `# Safety` section
  --> src/basic/string-util.rs:14:1
   |
14 | / pub unsafe extern "C" fn string_extract_line(s: *const c_char, i: usize, ret: *mut *mut c_char) -> c_int {
15 | |     let s = CStr::from_ptr(s).to_string_lossy();
16 | |     let mut lines = s.lines().enumerate().peekable();
17 | |
...  |
49 | |     return 0
50 | | }
   | |_^
   |
   = note: `#[warn(clippy::missing_safety_doc)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_safety_doc

warning: unneeded `return` statement
  --> src/basic/string-util.rs:49:5
   |
49 |     return 0
   |     ^^^^^^^^ help: remove `return`: `0`
   |
   = note: `#[warn(clippy::needless_return)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

warning: single-character string constant used as pattern
  --> src/basic/string-util.rs:22:39
   |
22 |             if i == 0 && !s.ends_with("\n") {
   |                                       ^^^^ help: try using a `char` instead: `'\n'`
   |
   = note: `#[warn(clippy::single_char_pattern)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_char_pattern

warning: 3 warnings emitted

```

Reply:
`rustfmt` isn't happy either
```diff
Diff in /home/vagrant/systemd/src/basic/string-util.rs at line 11:
 /// NULL, indicating the input string should be used as is, and avoid a memory allocation for a very
 /// common case.
 #[no_mangle]
-pub unsafe extern "C" fn string_extract_line(s: *const c_char, i: usize, ret: *mut *mut c_char) -> c_int {
+pub unsafe extern "C" fn string_extract_line(
+    s: *const c_char,
+    i: usize,
+    ret: *mut *mut c_char,
+) -> c_int {
     let s = CStr::from_ptr(s).to_string_lossy();
     let mut lines = s.lines().enumerate().peekable();

Diff in /home/vagrant/systemd/src/basic/string-util.rs at line 46:
         Err(_error) => return -1,
     };

-    return 0
+    return 0;
 }
```

Reply:
Thanks, fixed - unfortunately neither clippy nor rustfmt are packaged on Debian :-(

I don't think the Meson problem is related to the find_program, so I imagine there's some hard-coded assumption somewhere

Reply:
> unfortunately neither clippy nor rustfmt are packaged on Debian

I couldn't find any packages on Fedora either :-( Unfortunately it appears the official way to install those tools is with `rustup` (which in turn is installed with ["curl | sh"](https://rustup.rs/)) (which surprised me considering Rust is supposed to be some kind of safer alternative) (but I digress :-))

> I don't think the Meson problem is related to the find_program

Agreed

Reply:
Looks like clippy and fmt are still complaining:
```
$ /home/vagrant/.cargo/bin/clippy-driver -C linker=cc -C linker=cc --color=always --crate-type staticlib --crate-name basic_rust -g --emit dep-info=src/basic/basic_rust.d --emit link -o src/basic/libbasic_rust.a src/basic/string-util.rs
warning: unneeded `return` statement
  --> src/basic/string-util.rs:57:5
   |
57 |     return 0;
   |     ^^^^^^^^^ help: remove `return`: `0`
   |
   = note: `#[warn(clippy::needless_return)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

warning: 1 warning emitted

$ rustfmt --check src/basic/string-util.rs
Diff in /home/vagrant/systemd/src/basic/string-util.rs at line 16:
 /// allocated buffer that the caller needs to take ownership of.
 #[no_mangle]
 pub unsafe extern "C" fn string_extract_line(
-        s: *const c_char,
-        i: usize,
-        ret: *mut *mut c_char,
-    ) -> c_int {
+    s: *const c_char,
+    i: usize,
+    ret: *mut *mut c_char,
+) -> c_int {
     let s = CStr::from_ptr(s).to_string_lossy();
     let mut lines = s.lines().enumerate().peekable();
```

Reply:
mh the change was copied&pasted, now the indentation recommendation is different - weird

Reply:
@bluca looking at the CIFuzz failure (where `rustc` is installed in the base-builder image), I think the Rust code shouldn't be compiled unless `-Dbuild-rust=true` is set explicitly. It would help to avoid setting `-Dbuild-rust=false` everywhere. More generally, I don't think the Rust code (at this point and in the foreseeable future) should affect the systemd CI in the sense that it should just keep testing the C code as usual. If anyone is interested in testing the Rust code, they could add a new GHAction where ``-Dbuild-rust` could be set to `true`.

Reply:
> I couldn't find any packages on Fedora either :-(

There are: `clippy` and `rustfmt` packages.

Reply:
@bluca if you're interested there're 2 hopefully cherry-pickable commits at https://github.com/evverx/systemd/pull/5/commits/9d62cabda2388c96d8cbe5473ff86425d602d91f and https://github.com/evverx/systemd/pull/5/commits/42f686dd59d67211231c79ae1414caab64241ca7 that should make it possible to build the PR with ASan, clippy and rustfmt.

I kind of got rid of that YOLO "curl | sh" command the Rust site recommends by downloading `rustup` directly and verifying its checksum (which is a bit better but still ugly) but to be honest I'm not sure why they keep recommending it. I thought after  https://about.codecov.io/security-update/ they should have been urged by the Rust community to at least  start to slowly moving towards more sensible installation methods.

Reply:
yep noticed, funny because it's the style they suggest to use in the book :-) fixed now